File: util/makedeps.pl
#!/usr/bin/perl

my @all;

foreach $ARG (@ARGV) {
	open (FH, $ARG);

	my @depends;
	my @targets;

	$ARG =~ s:^\./::;

	push @depends, $ARG;
	
	($ARGDIR = $ARG) =~ s:[^/]*$::;
	
	while (<FH>) {
		if (/\bclass\s+(\w+)\s+extends\s+(\w+)\s*;/) {
			$class = $1;
			push @targets, "$ARGDIR$1.py";
		}
		if (/\bimportVocab\s*=\s*(\w+)\s*;/) {
			push @depends, "$ARGDIR$1TokenTypes.txt";
		}
		if (/\bexportVocab\s*=\s*(\w+)\s*;/) {
			$target = "$ARGDIR$1TokenTypes.txt";
			if (!grep {/$target/} @targets) {
				push @targets, $target;
			}
		}
	}

	print join(" ", @targets) . ': ' . join(" ", @depends) . "\n";
	print "\t\$(ANTLR) -o \$(\@D) \$<\n";
	print "\n";

	push @all, @targets;
}

print "all: " . join(" ", @all) . "\n\n";

print "clean: clean-antlr\n\n";

print "clean-antlr:\n";
print "\t\@\$(RM) " . join(" ", @all) . "\n\n";

File: util/makedeps.py
#!/usr/bin/env python

import sys
import re
import os.path

all = []

def enumerate_grammars():
	for dirpath, dirnames, filenames in os.walk('.'):
		for filename in filenames:
			if filename.endswith('.g'):
				yield os.path.join(dirpath, filename)
		if '.hg' in dirnames:
			dirnames.remove('.hg')

for arg in enumerate_grammars():
	arg = os.path.normpath(arg)
	argdir = os.path.dirname(arg)

	depends = []
	targets = []
	depends.append(arg)
	for line in open(arg, "rt"):
		mo = re.search(r"\bclass\s+(\w+)\s+extends\s+(\w+)\s*;", line)
		if mo:
			klass = mo.group(1)
			target = os.path.join(argdir, klass + ".py")
			targets.append(target)
		mo = re.search(r"\bimportVocab\s*=\s*(\w+)\s*;", line)
		if mo:
			depend = os.path.join(argdir, mo.group(1) + "TokenTypes.txt")
			depends.append(depend)
		mo = re.search(r"\bexportVocab\s*=\s*(\w+)\s*;", line)
		if mo:
			target = os.path.join(argdir, mo.group(1) + "TokenTypes.txt")
			if target not in targets:
				targets.append(target)

	print " ".join(targets) + ': ' + " ".join(depends)
	print "\t$(ANTLR) -o $(@D) $<"
	print "\t@touch " + " ".join(targets)
	print

	all.extend(targets)

print "all: " + " ".join(all)
print
print "clean: clean-antlr"
print
print "clean-antlr:";
print "\t-$(RM) " + " ". join(all)
print
